File: src/api/network/remoteJobs.ts
/**
 * Cross-device instruction dispatch IPC.
 *
 * Backing Rust: `src-tauri/src/commands/network/remote_jobs.rs`. The two reads
 * are plain DB reads and work with the network stopped (yesterday's history has
 * to be readable without the link being up); the send needs a live, paired,
 * reachable peer and fails typed when it is not.
 *
 * Like every other p2p surface, callers must clear `probeP2pSupport()` first —
 * see `@/lib/network/p2pCapability`. The store slice does that for the app; use
 * these directly only from code that has already probed.
 */
import { invokeWithTimeout as invoke } from '@/lib/tauriInvoke';
import type { RemoteJob } from '@/lib/bindings/RemoteJob';
import type { RemoteJobDirection } from '@/lib/bindings/RemoteJobDirection';
import type { RemoteJobNote } from '@/lib/bindings/RemoteJobNote';
import type { RemoteJobStatus } from '@/lib/bindings/RemoteJobStatus';

export type { RemoteJob, RemoteJobDirection, RemoteJobNote, RemoteJobStatus };

/**
 * The remote-job history, newest first. Omit `direction` for the merged
 * timeline; the backend treats `undefined` and `"all"` identically.
 */
export const listRemoteJobs = (direction?: RemoteJobDirection | 'all', limit?: number) =>
  invoke<RemoteJob[]>('list_remote_jobs', { direction, limit });

/** One job's progress notes, oldest first — the transcript under a job row. */
export const listRemoteJobNotes = (jobId: string) =>
  invoke<RemoteJobNote[]>('list_remote_job_notes', { jobId });

/**
 * Ask a paired device to run a natural-language instruction. Resolves once the
 * peer has acknowledged, so the returned row already says `running` (or carries
 * the refusal reason) without a second round trip.
 *
 * The wait is one LAN round trip, not the length of the turn — but it is a
 * round trip to a machine that may be mid-suspend, so it gets a longer leash
 * than the default before we call it a timeout.
 */
export const sendRemoteInstruction = (peerId: string, instruction: string, kind?: string) =>
  invoke<RemoteJob>('send_remote_instruction', { peerId, instruction, kind }, {
    timeoutMs: SEND_INSTRUCTION_TIMEOUT_MS,
  });

/** Ceiling on the peer's acknowledgement of a dispatched instruction. */
export const SEND_INSTRUCTION_TIMEOUT_MS = 45_000;

File: src-tauri/src/commands/network/remote_jobs.rs
//! Tauri commands for cross-device instruction dispatch.
//!
//! Thin wrappers over `engine::p2p::remote_jobs::RemoteJobs` (the wire and the
//! state machine) and the ungated `remote_jobs` repo (reads). Everything that
//! decides anything — the pairing gate, the typed offline refusal, the
//! exactly-once replay — lives in those two modules; nothing here is a policy
//! decision, so there is no second place for the trust check to drift to.

use tauri::State;

use crate::db::models::{RemoteJob, RemoteJobDirection, RemoteJobNote};
use crate::db::repos::resources::remote_jobs as repo;
use crate::error::AppError;
use crate::ipc_auth::{require_auth, require_auth_sync};
use crate::AppState;

fn jobs(
    state: &AppState,
) -> Result<&std::sync::Arc<crate::engine::p2p::remote_jobs::RemoteJobs>, AppError> {
    state
        .network
        .as_ref()
        .map(|net| &net.remote_jobs)
        .ok_or_else(|| AppError::Internal("Network service not initialized".into()))
}

/// The remote-job history, newest first. `direction` narrows to `"outbound"`
/// (work this device sent) or `"inbound"` (work it was asked to do); omit it for
/// both, which is how the combined timeline is read.
///
/// A pure DB read, so it works in a build with the network stopped — a user must
/// be able to look at what happened yesterday without the link being up.
#[tauri::command]
pub fn list_remote_jobs(
    state: State<'_, std::sync::Arc<AppState>>,
    direction: Option<String>,
    limit: Option<u32>,
) -> Result<Vec<RemoteJob>, AppError> {
    require_auth_sync(&state)?;
    let direction = match direction.as_deref() {
        None | Some("") | Some("all") => None,
        Some(d) => Some(RemoteJobDirection::parse(d).ok_or_else(|| {
            AppError::Validation(format!(
                "Unknown remote job direction '{d}' (expected 'outbound' or 'inbound')"
            ))
        })?),
    };
    repo::list(&state.db, direction, limit.unwrap_or(100))
}

/// One job's progress notes, oldest first — the transcript under a job row.
#[tauri::command]
pub fn list_remote_job_notes(
    state: State<'_, std::sync::Arc<AppState>>,
    job_id: String,
) -> Result<Vec<RemoteJobNote>, AppError> {
    require_auth_sync(&state)?;
    repo::list_notes(&state.db, job_id.trim())
}

/// Send a natural-language instruction to one of this user's paired devices and
/// wait for its acknowledgement.
///
/// Returns the persisted job once the peer has answered, so the caller can show
/// "running" or the refusal reason without a second round trip. Fails typed and
/// early when the device is not paired (`forbidden`) or not currently reachable
/// (`network_offline`, naming the device and the remedy) — neither leaves a row.
#[tauri::command]
pub async fn send_remote_instruction(
    state: State<'_, std::sync::Arc<AppState>>,
    peer_id: String,
    instruction: String,
    kind: Option<String>,
) -> Result<RemoteJob, AppError> {
    require_auth(&state).await?;
    jobs(&state)?
        .send_instruction(peer_id.trim(), kind, &instruction)
        .await
}
